### PR TITLE
[ES|QL] Add function log(base, value)

### PR DESCRIPTION
**Pull request by fang-xing-esql**
A scalar function log(based, value) returns the logarithm of a value for a particular base, as specified in the argument.

The arguments can be integer, long integer, unsigned long integer or double. If the arguments are not double, they are converted to one for processing by the function.

The result of the function is a double number.

The result can be null; if the argument is null, the result is the null value.

### CODE REVIEW

File: docs/reference/esql/functions/log.asciidoc [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+*Syntax*
+
+[.text-center]
+image::esql/functions/signature/log.svg[Embedded,opts=inline]
+
+`base`::
+Numeric expression. If `null`, the function returns `null`.
+
+`exponent`::
```
Comment (luigidellaquila) [inline docs/reference/esql/functions/log.asciidoc:13]: I think `exponent` is a bit confusing in the context of logarithm (the exponent is supposed to be the result...). Maybe `value` is good enough (and I see it's like that also in SHOW FUNCTIONS)

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/math.csv-spec [hunk 2]
```diff
@@ -209,6 +209,171 @@ salary:integer | s:double
 74572          | 193184
 73851          | 1330201
 ;
+// START To be removed
```
Comment (luigidellaquila) [inline x-pack/plugin/esql/qa/testFixtures/src/main/resources/math.csv-spec:212]: Leftover?

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/math/Log.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+import static org.elasticsearch.xpack.ql.expression.TypeResolutions.isNumeric;
+
+public class Log extends ScalarFunction implements OptionalArgument, EvaluatorMapper {
+
+    private final Expression base, value;
+
+    @FunctionInfo(returnType = "double", description = "Returns the logarithm of a value to a base.")
+    public Log(
+        Source source,
+        @Param(name = "base", type = { "integer", "unsigned_long", "long", "double" }) Expression base,
```
Comment (luigidellaquila) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/math/Log.java:40]: Maybe we could make the base optional and default to `10` or `e`.
See [DateParse ](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/date/DateParse.java) for an example

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/math/Log.java [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        return base.foldable() && value.foldable();
+    }
+
+    @Override
+    public Object fold() {
+        return EvaluatorMapper.super.fold();
+    }
+
+    @Evaluator(warnExceptions = { ArithmeticException.class })
+    static double process(double base, double value) {
```
Comment (luigidellaquila) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/math/Log.java:73]: In most of the cases the base will likely be a constant, so probably it makes sense to have a `process(double value)` method and a dedicated (and optimized) evaluator. See [this](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/date/DateParse.java#L103) as an example

**Review by luigidellaquila — COMMENTED** [review]
Thank you very much @fang-xing-esql, it looks pretty good.
Kudos for the test coverage.
I added a few comments, mostly for possible improvements

File: docs/reference/esql/functions/log.asciidoc (hunk 1 above)
Comment (fang-xing-esql) [reply docs/reference/esql/functions/log.asciidoc:13]: Thank you for catching the typo!

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/math.csv-spec (hunk 2 above)
Comment (fang-xing-esql) [reply x-pack/plugin/esql/qa/testFixtures/src/main/resources/math.csv-spec:212]: Removed.

**Review by astefan — COMMENTED** [review]
Please, add some csv-spec tests that use actual data from indices.
For example:
```
from employees | EVAL s = LOG(languages, salary) | keep emp_no, salary, s | limit 10
from employees | eval x = salary * 2 | eval s = log(x, x) | keep emp_no, salary, s | limit 10
from employees | rename languages as base | eval s = log(base, salary) | keep emp_no, salary, s | limit 10 | sort s nulls last
from employees | eval base = 2.5 | eval s = log(base, salary) * 0.5 | keep emp_no, salary, s | sort s nulls first | limit 5
```

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/math/Log.java (hunk 3 above)
Comment (fang-xing-esql) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/math/Log.java]: This is a good idea, default to base e if a base is not provided.

**Comment by fang-xing-esql** [comment]
> Please, add some csv-spec tests that use actual data from indices. For example:
> 
> ```
> from employees | EVAL s = LOG(languages, salary) | keep emp_no, salary, s | limit 10
> from employees | eval x = salary * 2 | eval s = log(x, x) | keep emp_no, salary, s | limit 10
> from employees | rename languages as base | eval s = log(base, salary) | keep emp_no, salary, s | limit 10 | sort s nulls last
> from employees | eval base = 2.5 | eval s = log(base, salary) * 0.5 | keep emp_no, salary, s | sort s nulls first | limit 5
> ```

More csv tests on indices are added.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/math/Log.java (hunk 4 above)
Comment (fang-xing-esql) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/math/Log.java]: Discussed with Luigi, this enhancement can be deferred to a follow-up PR.

**Review by astefan — APPROVED** [review]
Left a small cosmetic comment, otherwise LGTM. Well done!

**Review by luigidellaquila — APPROVED** [review]
LGTM, thanks!

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/math.csv-spec [hunk 5]
```diff
@@ -210,6 +210,307 @@ salary:integer | s:double
...
+
+logofNegativeZero#[skip:-8.12.99,reason:new scalar function added in 8.13]
+row base = 2.0, value = -0.0
+| EVAL s = LOG(base, value);
+
+warning:Line 2:12: evaluation of [LOG(base, value)] failed, treating result as null. Only first 20 failures recorded.
+warning:Line 2:12: java.lang.ArithmeticException: Log of non-positive number
+
+base:double | value:double | s:double
+2.0         | -0.0             |null
```
[the remark this answers is not in the document]
Comment (fang-xing-esql) [reply x-pack/plugin/esql/qa/testFixtures/src/main/resources/math.csv-spec]: Fixed in the latest commit.

**Pull request merged**